### PR TITLE
fix: Rename `Protocol::WebRTC` to `Protocol::WebRTCDirect`

### PR DESCRIPTION
**Pull request by mxinden**
See https://github.com/multiformats/multiaddr/pull/150#issuecomment-1468791586 for context.

@melekes @thomaseizinger can you give this a review?

### CODE REVIEW

File: src/protocol.rs [hunk 1]
```diff
@@ -202,7 +202,7 @@ impl<'a> Protocol<'a> {
             }
             "p2p-websocket-star" => Ok(Protocol::P2pWebSocketStar),
             "p2p-webrtc-star" => Ok(Protocol::P2pWebRtcStar),
-            "webrtc" => Ok(Protocol::WebRTC),
+            "webrtc-direct" => Ok(Protocol::WebRTCDirect),
```
Comment (melekes) [inline src/protocol.rs:205]: I must say having both `webrtc-direct` and `p2p-webrtc-direct` is confusing, but I guess nothing we can do about it other than good documentation.

**Review by melekes — APPROVED** [review]
👍

File: src/protocol.rs (hunk 1 above)
Comment (mxinden) [reply src/protocol.rs:205]: We could as well deprecate `p2p-webrtc-direct`. I am not aware of anyone using it. Though intuitively, that decision should happen on https://github.com/multiformats/multiaddr.

**Review by thomaseizinger — COMMENTED** [review]
Could we first release a version where we accept parsing as `/webrtc-direct` and not rename the variant, thus making this a non-breaking change?

I'd like to batch some breaking changes in `rust-multiaddr` together, e.g. https://github.com/multiformats/rust-multiaddr/pull/77 and https://github.com/multiformats/rust-multiaddr/issues/73 but those are not ready yet. Especially https://github.com/multiformats/rust-multiaddr/issues/73 needs a breaking change in `libp2p-identity` first which implies a round of breaking changes to `libp2p`.

**Comment by mxinden** [comment]
> Could we first release a version where we accept parsing as `/webrtc-direct` and not rename the variant, thus making this a non-breaking change?

On first thought this is a great idea. Though I am not sure there are no hidden surprises with this strategy. E.g. the guarantee of round-trip parsing (`/webrtc-direct` -> binary -> `/webrtc-direct`) no longer holds.

Breaking change:
- Avoid long-term confusion, i.e. get it over with quickly
- Needs a breaking change in `libp2p`

Non-breaking change:
- Safes us and our users a lot of work
- Might be a footgun
- Could also default to `/webrtc-direct` from binary to text already now
- Takes time to fully finish, thus less time between now and the introduction of browser-to-browser `/webrtc`

Thoughts @thomaseizinger?

**Comment by mxinden** [comment]
Friendly ping @thomaseizinger. What are your thoughts here?

**Comment by thomaseizinger** [comment]
> Friendly ping @thomaseizinger. What are your thoughts here?

Didn't submit my comment :man_facepalming:

**Comment by thomaseizinger** [comment]
> > Could we first release a version where we accept parsing as `/webrtc-direct` and not rename the variant, thus making this a non-breaking change?
> 
> On first thought this is a great idea. Though I am not sure there are no hidden surprises with this strategy. E.g. the guarantee of round-trip parsing (`/webrtc-direct` -> binary -> `/webrtc-direct`) no longer holds.
> Breaking change:
> - Avoid long-term confusion, i.e. get it over with quickly
> - Needs a breaking change in `libp2p`
> 
> Non-breaking change:
> - Safes us and our users a lot of work
> - Might be a footgun
> - Could also default to `/webrtc-direct` from binary to text already now
> - Takes time to fully finish, thus less time between now and the introduction of browser-to-browser `/webrtc`
> 
> Thoughts @thomaseizinger?

What we'd need to work out is how exactly we are going to introduce the other `/webrtc` protocol after this.

We definitely need to log a deprecation warning if we are now adding the 2nd representation.

Then in a breaking change, we can rename the variant and add the "new" `/webrtc` protocol.

How big is the footgun actually?

**Comment by thomaseizinger** [comment]
> How big is the footgun actually?

The only problem I can think of is that in the "breaking" change where we add the new (browser-to-browser) `/webrtc` protocol, users of `multiaddr` might not realise that the `WebRTC` protocol semantically changes. However, afaik the only user of this is us in `libp2p-webrtc` and we know to be careful here.

Ideally, semantic breaking changes are also compile-time breaking changes but I'd say with sufficient documentation, we can also deviate from that. I've pushed https://github.com/multiformats/rust-multihash/pull/272 forward a lot so I'd love if the next breaking change can already include that (and thus be the last one in a while).

**Comment by thomaseizinger** [comment]
If we move forward with this, it would have to be done as a patch-release from `0.17.0` because `master` already contains breaking changes.

**Comment by mxinden** [comment]
Pull request for the non-breaking change: https://github.com/multiformats/rust-multiaddr/pull/84

**Comment by mxinden** [comment]
With https://github.com/multiformats/rust-multiaddr/pull/84 released as `v0.17.1`, I updated this pull request, making the breaking change of renaming `Protocol::WebRTC` to `Protocol::WebRTCDirect` and removal of the deprecated `/webrtc` string representation.

**Comment by mxinden** [comment]
I seem to only be able to request a review from one of the two of you.

**Review by thomaseizinger — APPROVED** [review]

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,9 +1,15 @@
-# 0.18.0 [unreleased]
+# 0.18.0 - unreleased
 
 - Add `WebTransport` instance for `Multiaddr`. See [PR 70].
 - Disable all features of `multihash`. See [PR 77].
 - Mark `Protocol` as `#[non_exhaustive]`. See [PR 82].
```
[the remark this answers is not in the document]
Comment (mxinden) [reply CHANGELOG.md:5]: Correct

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,9 +1,15 @@
...
 
+- Rename `Protocol::WebRTC` to `Protocol::WebRTC`.
```
[the remark this answers is not in the document]
Comment (mxinden) [reply CHANGELOG.md:7]: :pray: uups

**Comment by DougAnderson444** [comment]
@mxinden when is this v0.18 planned for release so that we can use the new naming without patches?

**Comment by thomaseizinger** [comment]
> I'd first like to get a release of `rust-multihash` out: [multiformats/rust-multihash/milestone/1](https://github.com/multiformats/rust-multihash/milestone/1)

I've created a milestone for `rust-multiaddr` too so we can easily keep track of this: https://github.com/multiformats/rust-multiaddr/milestone/3

**Pull request merged**